modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1925,29 +1922,41 @@ void GDScriptAnalyzer::resolve_return(GDScriptParser::ReturnNode *p_return) {
 		}
 		result = p_return->return_value->get_datatype();
 	} else {
-		// Return type is null by default.
 		result.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
 		result.kind = GDScriptParser::DataType::BUILTIN;
 		result.builtin_type = Variant::NIL;
 		result.is_constant = true;
 	}
 
 	if (has_expected_type) {
-		expected_type.is_meta_type = false;
-		if (expected_type.is_hard_type()) {
-			if (!is_type_compatible(expected_type, result)) {
-				// Try other way. Okay but not safe.
+		if (expected_type.is_variant()) {
+			// variant expectation
```

Comment:
This is style nitpick but ideally comments should be in sentence case and end with a period.

Reply:
Yeah, personally I do not like comments, I put them here to explicitly show what is current condition because there are a lot of `if-else`. And those comments made to match same in `reduce_assignment` with the idea that the logic can be moved in a separate function later.

Reply:
Done. Removed comments.